packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -0,0 +1,136 @@
+const stateMetadata = {
+  queue: { persist: false, anonymous: false },
+};
+
+const getDefaultState = () => ({});
+
+export const QueuedRequestControllerActionTypes = {
+  enqueueRequest: `${controllerName}:enqueueRequest` as const,
+};
+
+export const QueuedRequestControllerEventTypes = {
+  stateChange: `${controllerName}:stateChange` as const,
+  countChanged: `${controllerName}:countChanged` as const,
+};
+
+export type QueuedRequestControllerState = {
+  [k: string]: any;
+};
+
+export type QueuedRequestControllerStateChangeEvent = {
+  type: typeof QueuedRequestControllerEventTypes.stateChange;
+  payload: [QueuedRequestControllerState, Patch[]];
+};
+
+export type QueuedRequestControllerCountChangedEvent = {
+  type: typeof QueuedRequestControllerEventTypes.countChanged;
+  payload: [number];
+};
+
+export type QueuedRequestControllerEnqueueRequestAction = {
+  type: typeof QueuedRequestControllerActionTypes.enqueueRequest;
+  handler: QueuedRequestController['enqueueRequest'];
+};
+
+export type QueuedRequestControllerEvents =
+  | QueuedRequestControllerStateChangeEvent
+  | QueuedRequestControllerCountChangedEvent;
+
+export type QueuedRequestControllerActions =
+  QueuedRequestControllerEnqueueRequestAction;
+
+export type QueuedRequestControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  QueuedRequestControllerActions,
+  QueuedRequestControllerEvents,
+  string,
+  string
+>;
+
+export type QueuedRequestControllerOptions = {
+  messenger: QueuedRequestControllerMessenger;
+};
+
+/**
+ * Controller for request queueing.
+ */
+export class QueuedRequestController extends BaseControllerV2<
+  typeof controllerName,
+  QueuedRequestControllerState,
+  QueuedRequestControllerMessenger
+> {
+  private currentRequest: Promise<unknown> = Promise.resolve();
+
+  #count = 0;
+
+  /**
+   * Construct a RequestQueueController.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the QueuedRequestController
+   */
+  constructor({ messenger }: QueuedRequestControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+    this.#registerMessageHandlers();
+  }
+
+  #registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      this.enqueueRequest.bind(this),
+    );
+  }
+
+  length() {
+    return this.#count;
+  }
+
+  // [ current batch ] - [ batch n ] - [ last batch ]
+  // for new request
+  // if origin is not the same as last batch origin
+  // make new batch / enqueueRequest
+  // otherwise, add request to the last batch
+
+  async enqueueRequest(requestNext: (...arg: unknown[]) => Promise<unknown>) {
+    this.#count += 1;
+    this.messagingSystem.publish(
+      'QueuedRequestController:countChanged',
+      this.#count,
+    );
```

Comment:
seeing code shaped this way 3 times in this func. Is it worth DRYing into a helper?